src/main/flight/gtune.c

```diff
@@ -0,0 +1,210 @@
+    pidProfile->gtune_hilimP[YAW]    = 100; [0..200] Higher limit of YAW P during G tune. 0 Disables tuning for that axisis.
+    pidProfile->gtune_pwr            = 0;   [0..10] Strength of adjustment
+    pidProfile->gtune_settle_time    = 450; [200..1000] Settle time in ms
+    pidProfile->gtune_average_cycles = 16;  [8..128] Number of looptime cycles used for gyro average calculation
+*/
+
+static pidProfile_t *pidProfile;
+static int16_t delay_cycles;
+static int16_t time_skip[3];
+static int16_t OldError[3], result_P64[3];
+static int32_t AvgGyro[3];
+static bool floatPID;
+
+void updateDelayCycles(void)
+{
+    delay_cycles = -(((int32_t)pidProfile->gtune_settle_time * 1000) / cycleTime);
+}
+
+void init_Gtune(pidProfile_t *pidProfileToTune)
+{
+    uint8_t i;
+
+    pidProfile = pidProfileToTune;
+	if (pidProfile->pidController == 2) {
+	    floatPID = true;                                                        // LuxFloat is using float values for PID settings
+	} else {
+	    floatPID = false;
+	}
+	updateDelayCycles();
+	for (i = 0; i < 3; i++) {
+        if ((pidProfile->gtune_hilimP[i] && pidProfile->gtune_lolimP[i] > pidProfile->gtune_hilimP[i]) || (motorCount < 4 && i == FD_YAW)) { // User config error disable axisis for tuning
+            pidProfile->gtune_hilimP[i] = 0;                                    // Disable YAW tuning for everything below a quadcopter
+        }
+        if (floatPID) {
+            if((pidProfile->P_f[i] * 10.0f) < pidProfile->gtune_lolimP[i]) {
+                pidProfile->P_f[i] = (float)(pidProfile->gtune_lolimP[i] / 10.0f);
+            }
+            result_P64[i] = (int16_t)pidProfile->P_f[i] << 6;                   // 6 bit extra resolution for P.
+        } else {
+            if(pidProfile->P8[i] < pidProfile->gtune_lolimP[i]) {
+                pidProfile->P8[i] = pidProfile->gtune_lolimP[i];
+            }
+            result_P64[i] = (int16_t)pidProfile->P8[i] << 6;                    // 6 bit extra resolution for P.
+        }
+        OldError[i] = 0;
+        time_skip[i] = delay_cycles;
+    }
+}
+
+void calculate_Gtune(uint8_t axis)
+{
+    int16_t error, diff_G, threshP;
+
+    if(rcCommand[axis] || (axis != FD_YAW && (FLIGHT_MODE(ANGLE_MODE) || FLIGHT_MODE(HORIZON_MODE)))) {  // Block tuning on stick input. Always allow G-Tune on YAW, Roll & Pitch only in acromode
+        OldError[axis] = 0;
+        time_skip[axis] = delay_cycles;                                         // Some settle time after stick center. default 450ms
+    } else {
+        if (!time_skip[axis]) AvgGyro[axis] = 0;
+        time_skip[axis]++;
+        if (time_skip[axis] > 0) {
+            if (axis == FD_YAW) {
+                AvgGyro[axis] += 32 * ((int16_t)gyroData[axis] / 32);           // Chop some jitter and average
+            } else {
+                AvgGyro[axis] += 128 * ((int16_t)gyroData[axis] / 128);         // Chop some jitter and average

```

Comment:
(from @Spirre logs):
This quantization is generating zero AvgGyro on most cycles unless there is external disturbance. Step in PID tuning causes enough quad vibration to sustain measurement for some time.

`thresP` is only 10 - less than 10% of quantization